### PR TITLE
Add typing to the Playlist and Page classes

### PR DESCRIPTION
**Pull request by arusahni**
This was the hairiest of the type conversions so far. It was complicated by the fact that the `Artist` class does not have an items collection. As a result, the `Page` iterator interface would end up raising an `AttributeError` if iteration was attempted. Rather than add a no-op `items` method to the the `Artist` type, I opted to shim it at the `Page` level to keep the API clean.

Please nitpick this! While I use the API, I've never used this module.

Advances #137

### CODE REVIEW

**Comment by tehkillerbee** [comment]
Thanks, I will try to test this asap, as I just added a few features to mopidy-tidal that requires Page objects.

**Comment by 2e0byo** [comment]
Hmm.  Thanks for this.

I think you've exposed a real weakness in the class hierarchy here.  I've not had a massive amount of thought about it, but instinctively I'd prefer making the spec of all the response types the same (even with a no-op .items()), preferably by making them inherit and implement an ABC.  For the time being @arusahni could you do it the other way, i.e. change the class?

As a more general point I'd like to have a shot at thinning out the codebase a bit.  The api as it stands is quite fragmented, something which lead to some curious mocking tricks when I first wrote the test suite for mopidy-tidal. Getting typing (and testing) working is a good way to avoid nasty surprises, but I'm personally up for changing our api (with careful consideration) to get rid of some of these inconsistencies.  As a step in that direction I think it would be good if PRs started proposing api improvements, i.e. don't feel afraid @arusahni to suggest changes.

At some point we need to rewrite the config class at least a bit.

I don't have massive amounts of time before Christmas, but I might manage to get some work done on things in december.  If anyone's browsing by and want to have a shot at refactoring, go for it.

File: tidalapi/page.py [hunk 1]
```diff
@@ -19,54 +19,100 @@
...
     A page from the https://listen.tidal.com/view/pages/ endpoint
 
     The :class:`categories` field will the most complete information
     However it is an iterable that goes through all the visible items on the page as well, in the natural reading order
     """
 
-    title = ""
-    categories: Optional[List[Any]] = None
-    _categories_iter: Optional[Iterator[Any]] = None
+    title: Optional[str] = ""
```
Comment (2e0byo) [inline tidalapi/page.py:72]: Is it ever valid for the title to be unset?  why not just str?

Also more generally, do we *need* a default for these?  Lots of this code was written before typing exists, when the pattern was to set an empty sentinel on the class to indicate the spec.

See also https://github.com/tamland/python-tidal/issues/192 .  That's a bigger issue than here and I don't mind leaving the rest as optional for now, but I think this should just be `str`, as it *defaults* to a str.

File: tidalapi/page.py [hunk 2]
```diff
@@ -19,54 +19,100 @@
...
+    categories: Optional[List["AllCategories"]] = None
+    _categories_iter: Optional[Iterator["AllCategories"]] = None
+    _items_iter: Optional[Iterator[Callable[..., Any]]] = None
+    page_category: "PageCategory"
+    request: "Requests"
 
-    def __init__(self, session, title):
+    def __init__(self, session: "Session", title: Optional[str]):
```
Comment (2e0byo) [inline tidalapi/page.py:79]: following on I'd just make that `title: str`.

File: tidalapi/page.py [hunk 3]
```diff
@@ -19,54 +19,100 @@
...
         self.request = session.request
         self.categories = None
         self.title = title
         self.page_category = PageCategory(session)
 
-    def __iter__(self):
+    def __iter__(self) -> "Page":
```
Comment (2e0byo) [inline tidalapi/page.py:85]: (aside: when we move to 3.11 these should become `typing.Self` to handle children.)

File: tidalapi/page.py [hunk 4]
```diff
@@ -99,17 +145,30 @@ def get(self, endpoint: str, params: Optional[Dict[str, Any]] = None) -> "Page":
...
 
 
-class PageCategory(object):
+@dataclass
+class More:
+    api_path: str
+    title: str
+
+    @staticmethod
+    def from_response(json_obj: JsonObj) -> Optional["More"]:
```
Comment (2e0byo) [inline tidalapi/page.py:154]: I prefer your naming convention, but the rest of the library uses `parse`, so I think we should here.

Comment (2e0byo) [reply tidalapi/page.py:154]: Also why not a class method and then return `cls()` at the end?

File: tidalapi/page.py [hunk 5]
```diff
@@ -99,17 +145,30 @@ def get(self, endpoint: str, params: Optional[Dict[str, Any]] = None) -> "Page":
...
+        show_more = json_obj.get("showMore")
+        if show_more is None:
+            return None
+        return More(api_path=show_more["apiPath"], title=show_more["title"])
```
Comment (2e0byo) [inline tidalapi/page.py:158]: nit: I prefer an explicit `else:` (one branch per return type is easier to grok imho)

File: tidalapi/page.py [hunk 6]
```diff
@@ -118,25 +177,23 @@ def __init__(self, session: "tidalapi.session.Session"):
...
-                PageLinks, FeaturedItems, ItemList, TextBlock, LinkList
-            ] = PageLinks(self.session)
+            category: PageCategories = PageLinks(self.session)
         elif category_type in ("FEATURED_PROMOTIONS", "MULTIPLE_TOP_PROMOTIONS"):
             category = FeaturedItems(self.session)
         elif category_type in self.item_types.keys():
             category = ItemList(self.session)
         elif category_type == "MIX_HEADER":
-            return self.session.parse_mix(json_obj["mix"])
+            return cast("Mix", self.session.parse_mix(json_obj["mix"]))
```
Comment (2e0byo) [inline tidalapi/page.py:190]: why do we need the cast?  Can't we just typehint `parse_mix`?

File: tidalapi/page.py [hunk 7]
```diff
@@ -179,12 +230,12 @@ def show_more(self):
 class FeaturedItems(PageCategory):
     """Items that have been featured by TIDAL."""
 
-    items: Optional[list["PageItem"]] = None
+    items: Optional[List["PageItem"]] = None
 
-    def __init__(self, session):
+    def __init__(self, session: "Session"):
         super(FeaturedItems, self).__init__(session)
```
Comment (2e0byo) [inline tidalapi/page.py:236]: not in your pr but should be `super().__init__(session)`. (this is just old code)

File: tidalapi/page.py [hunk 8]
```diff
@@ -305,37 +359,34 @@ def __init__(self, session, json_obj):
...
+    text: str = ""
+    icon: str = ""
+    items: Optional[List[str]] = None
 
-    def __init__(self, session):
+    def __init__(self, session: "Session"):
         self.session = session
 
-    def parse(self, json_obj):
+    def parse(self, json_obj: JsonObj) -> "TextBlock":
```
Comment (2e0byo) [inline tidalapi/page.py:389]: not your code but I'm just reading this thoroughly for the first time.. This should be a staticmethod/classmethod.  I think I'll make a general refactoring issue.

File: tidalapi/request.py [hunk 9]
```diff
@@ -143,15 +163,22 @@ def map_json(cls, json_obj, parse=None, session=None):
                 for item in items:
                     item["item"]["dateAdded"] = item["created"]
 
-            lists = []
+            lists: List[Any] = []
             for item in items:
                 if session is not None:
-                    parse = session.convert_type(
-                        item["type"].lower() + "s", output="parse"
+                    parse = cast(
```
Comment (2e0byo) [inline tidalapi/request.py:169]: this general logic is *definitely* a code smell.  Leave it for now though.

**Review by 2e0byo — CHANGES_REQUESTED** [review]
Thanks for this. A few nits + the inverse shimming option, but otherwise LGTM.

File: tidalapi/page.py (hunk 8 above)
Comment (2e0byo) [reply tidalapi/page.py:389]: See #193

File: tidalapi/page.py (hunk 1 above)
Comment (arusahni) [reply tidalapi/page.py:72]: In [session.py](https://github.com/tamland/python-tidal/blob/master/tidalapi/session.py#L233) we initialize a `Page` with `None` as the `title`, however. If we're fine swapping that to an empty string, I'm fine stripping out the `Optional` here.

File: tidalapi/page.py (hunk 4 above)
Comment (arusahni) [reply tidalapi/page.py:154]: A class method would be fine here, too. I'll make these changes.

File: tidalapi/page.py (hunk 6 above)
Comment (arusahni) [reply tidalapi/page.py:190]: This line causes mypy to fall over in strict mode, so I added the explicit casts to quell some of the errors.

> tidalapi/page.py:191: error: Returning Any from function declared to return "Artist | Album | PageLinks | FeaturedItems | ItemList | TextBlock | LinkList | Mix"  [no-any-return]

I think cleaning up `Session` will help improve the downstream inferences.

File: tidalapi/page.py (hunk 7 above)
Comment (arusahni) [reply tidalapi/page.py:236]: Yeah, will make the change since I've been stripping out new-style class declarations anyway.

File: tidalapi/page.py (hunk 8 above)
Comment (arusahni) [reply tidalapi/page.py:389]: Yeah, a lot of our typings could be simplified if we move away from the initialize-as-empty-then-hydrate-with-a-response-and-copy technique.

File: tidalapi/page.py (hunk 6 above)
Comment (2e0byo) [reply tidalapi/page.py:190]: makes sense; leave for now then.  casts are a big code smell so they'll draw attention to it.

File: tidalapi/page.py (hunk 1 above)
Comment (2e0byo) [reply tidalapi/page.py]: I'd missed that.  don't really like it [requires null checking in everything downstream] but it's an implementation problem then, so keep it as optional.

File: tidalapi/page.py (hunk 6 above)
Comment (2e0byo) [reply tidalapi/page.py:191]: Actually, looks like you can do it in session.

in the constructor:

```python
        self.parse_mix: Callable[..., mix.Mix] = self.mix().parse
```

And then drop the casts here.

I'd prefer that as it's providing the information where it's needed.  (I'd prefer not dynamically constructing these methods but that can wait unless you'd rather rewrite them as:

```python
def parse_mix(self, obj: dict) -> Mix:
    return self.mix().parse(obj)
```

Which I guess is rather verbose.)

Separately, how are you testing?  `mypy --strict tidalapi/page.py` throws 5 errors for me.

File: tidalapi/page.py (hunk 5 above)
Comment (arusahni) [reply tidalapi/page.py]: Done.

File: tidalapi/page.py (hunk 6 above)
Comment (arusahni) [reply tidalapi/page.py]: > Actually, looks like you can do it in session.

Unfortunately that doesn't work. It should, but mypy is giving up somewhere. Refactoring them into standalone methods worked. Along the way, I discovered an issue with some of the earlier session typings I introduced: I was confused by/following the docstrings and assumed we re-exported the models from the root module. We don't :facepalm: . I reparented the references to point to the relevant model modules.

**Comment by arusahni** [comment]
@2e0byo thanks for the review! I sprinkled a bunch of `reveal_types` for every item in a chained call, and found two core issues that broke mypy's inference (pyright was fine):

1. Broken typings on `Session`
2. Missing typings for the `playlist` module.

I've updated this PR to include typings for `playlist`, and fixed #1. There are now far fewer casts.

**Comment by tehkillerbee** [comment]
@arusahni I did some quick tests in Mopidy-Tidal. It looks like everything is working as expected when browsing various pages so at least it looks like these changes didn't break anything.

**Comment by 2e0byo** [comment]
Thanks @arusahni yet again for the work.  I'll run the test suite at some point and then merge this in.

**Comment by tehkillerbee** [comment]
Any more comments on this PR @2e0byo? Otherwise lets merge it in and prepare for next tidalapi release.

**Comment by 2e0byo** [comment]
@tehkillerbee if the tests pass all is good.  Would you check?  Thanks for picking this up.

File: tidalapi/playlist.py [hunk 10]
```diff
@@ -20,38 +20,45 @@
...
+    description: Optional[str] = None
+    duration: int = -1
+    last_updated: Optional[datetime] = None
+    created: Optional[datetime] = None
     type = None
     public: Optional[bool] = False
-    popularity = -1
-    promoted_artists = None
-    last_item_added_at = None
+    popularity: Optional[int] = -1
```
Comment (tehkillerbee) [inline tidalapi/playlist.py:53]: We should probably set this one to None as well

File: tidalapi/playlist.py [hunk 11]
```diff
@@ -83,7 +90,8 @@ def parse(self, json_obj):
         self.created = dateutil.parser.isoparse(created) if created else None
         public = json_obj.get("publicPlaylist")
         self.public = None if public is None else bool(public)
-        self.popularity = json_obj.get("popularity")
+        popularity = json_obj.get("popularity")
+        self.popularity = int(popularity) if popularity else None
```
Comment (tehkillerbee) [inline tidalapi/playlist.py:94]: This breaks the tests, as we expect default value to be 0. However, your changes are more consistent so lets update the tests to expect None

**Comment by tehkillerbee** [comment]
LGTM, will fix remaining comments in a separate PR
